README
pam_pgsql 0.7
=============

Introduction
============

This module provides support to authenticate against PostgreSQL
tables for PAM-enabled appliations.

This module is based in part on the FreeBSD pam_unix module, and
the Debian pam_mysql module, but was written from scratch using
the two as a reference.

There is another pam_pgsql module, but the sources appear to have
vanished, hence this module.

Changes since last release
==========================

See the file CHANGELOG.

Compilation & Installation
==========================

pam_pgsql is now autoconf'ed, thus, compiling should be a matter
of:

    $ ./configure
    $ make
    $ make install

Or if you're using a git version, run this command before them all:

    $ ./autogen.sh

Compilation has been tested on Debian GNU/Linux, ArchLinux and FreeBSD 7.2

On Debian, you will need the libpam0g-dev and postgresql-dev packages to compile.
On FreeBSD you will have to install the postgresql/postgresql8*-client port.

See authenticate.c and chpass.c for an example application that authenticates and change password using this module.

This version only works with PostgreSQL versions 7.4 or newer.

Configuration
=============

For the service you wish the module to be used, you need
to edit the /etc/pam.d/<service> file or /etc/pam.conf, and 
add the relevant lines.

For example:

auth        required    pam_pgsql.so 
account     required    pam_pgsql.so
password    required    pam_pgsql.so
session     required    pam_pgsql.so

Or:

password    required    pam_cracklib.so
password    required    pam_pgsql.so authtok

Configure the database, and table the module should use with
the configuration file /etc/pam_pgsql.conf. An example of
this file:

database = sysdb
user = ljb
table = account
user_column = user_name
pwd_column = user_password
expired_column = acc_expired
newtok_column = acc_new_pwreq
debug = 1

Note that for backwards compatibility with earlier versions, options specified
in the configuration file can be supplied as module arguments as well. Module
arguments will override the configuration file.

From version 0.6 you can also use new style configuration (overrides legacy
values).

connect = dbname=sysdb user=ljb password=sth connect_timeout=15
auth_query = select user_password from account where user_name = %u
acct_query = select (acc_expired = 'y' OR acc_expired = '1'), (acc_new_pwreq = 'y' OR acc_new_pwreq = '1'), (user_password IS NULL OR user_password = '') from account where user_name = %u
pwd_query = update account set user_password = %p where user_name = %u


Configuration Options
=====================

    connect 		- the database connection string (see http://www.postgresql.org/docs/7.4/interactive/libpq.html#LIBPQ-CONNECT)
			  overrides other connection specific options
    auth_query		- authentication query (should return one column -- password)
			  overrides other authentication specific options
    auth_succ_query	- query to be executed after successful authentication
    auth_fail_query	- query to be executed after failed authentication
    session_open_query  - query to be executed on session start
    session_close_query - query to be executed on session end

    acct_query		- account options query (should return 3 or 4 boolean columns -- expired, new password required and password is null, return PAM_PERM_DENIED (y/n))
			  overrides other account specific options
    pwd_query		- query to be executed for password changing 
			  overrides other settings related to changing password

You can use %u as username, %p as (new) password, %h for hostname of client
as specified by PAM subsystem, %i for IP got by gethostbyname(%h) and %s as
pa service name in any query. Please don't forget to specify pw_type as %p
is replaced by password of pw_type form.

Caution: 
If %h is unavailable but used, system substitutes it in query with
NULL, but does not fail (you can fail it manually by using "%s is not null"
somewhere in your query). If %i is used and gethostbyname() fails than:
	(i) when rhost is empty or doesn't contain any periods ("."), %i is
replaced with NULL
	(ii) in any other case pam_pgsql return PAM_AUTH_ERR. 

    database            - the database which should be connected to
    table               - the table containing the authentication data
    host		- the host database server is running on (leave empty for socket)
    port		- the port database server is running on (leave empty for socket)
    user                - the username used when connecting to PostgreSQL
    password            - the password for the user specified
    user_column         - the column containing usernames
    pwd_column          - the column containing the passwords
    expired_column      - this column should contain '1' or 'y' if the account
                          has expired, bool type is OK
    newtok_column       - this column should contain '1' or 'y' if the user
                          needs to change their password, bool type is OK
    debug               - this is a standard module option that will enable
                          debug output to syslog (takes no values)
    pw_type             - specifies the password encryption scheme, can be one
                          of 'clear', 'md5', 'sha1', 'crypt', 'crypt_md5', or
						  'md5_postgres'. The difference between 'md5' and
						  'crypt_md5' is that 'md5' uses libmhash for hashing
						  while 'crypt_md5' uses crypt() with a special salt to
						  select md5 hashing instead of DES. if one of 'crypt'
						  or 'crypt_md5' is specified, passwords always are
                          encrypted in the respective format. however,
                          passwords in both formats may be stored in the
                          database, just as with /etc/(passwd|shadow).
                          defaults to 'clear'.
						  'md5_postgres' uses the postgres default internal
						  algorithm where hash is md5||md5(password+login). This
						  is usefull for authenticating against postgres users
						  created by the createuser postgres command.
    config_file         - alternative location of configuration file - it should be
			  specified as module argument.
    timeout		- if specified pam-pgsql will wait for timeout
			  seconds before giving up on db connection

There are also additional flags you can use:
    authtok		- see "use_first_pass"
    use_first_pass	- require authtok from previous entry in PAM stack
			  (useful for "password   required  pam_pgsql.so authtok")
			  after "password required pam_cracklib.so ..."
    try_first_pass	- same as previous, but doesn't fail if previous
			  module failed to provide us with password
    echo_pass 		- displays password while being typed

Example to autenticate against postgres users
=============================================
database = postgres
user = postgres
table = pg_catalog.pg_shadow
user_column = usename
pwd_column = passwd
pw_type=md5_postgres


CHANGELOG
pam-pgsql (0.7.3.2) experimental; urgency=low

  * add an "exception" to the GPL license to allow linking to non-GPL code
  * don't use access(2) but check error from fopen(2) when reading the
    configuration file
  * Fix support for pw_type md5_postgres
  * Fix NULL password query result permits login with any password

 -- William Grzybowski <wg@FreeBSD.org>  Thu, 02 Oct 2014 09:15:36 -0300

pam-pgsql (0.7.3.1) experimental; urgency=low

  * Update CHANGELOG
  * Bump autoconf version
  * include autogen.sh in autotools generated tarball

 -- Jan Dittberner <jandd@debian.org>  Sun, 27 Mar 2011 08:47:48 +0200

pam-pgsql (0.7.3) experimental; urgency=low

  * Fix for PostgreSQL authentication by Nirgal Vourgère (Debian Bug #594721)
  * Better logging for failed login attempt by Jan Dittberner

 -- William Grzybowski <william@agencialivre.com.br>  Thu, 03 Mar 2011 14:54:19 +0000

pam-pgsql (0.7.2) experimental; urgency=high

  * Bugfix in address to string conversion

 -- William Grzybowski <william@agencialivre.com.br>  Thu, 03 Mar 2011 14:28:17 +0000

pam-pgsql (0.7.1) experimental; urgency=low

  * Autotools refactor
  * Visibility ELF support
  * Linux-PAM vs OpenPAM checks (no more OS guessing)
  * Several bug fixes

  * Special thanks to Diego Elio “Flameeyes” Pettenò, Gentoo developer

 -- William Grzybowski <william@agencialivre.com.br>  Tue, 15 Mar 2010 10:10:00 -0300

pam-pgsql (0.7) experimental; urgency=low

  * Internal code rework
    - spliting the PostgreSQL code from PAM
    - spliting the module options code from the PAM
  * SHA1 password support
  * Critical module options are no longer avaible from pam.d/pam.conf for security issues
  * Sslmode module option was added and is recommended to use as "require" when the postgresql database is not localhost to prevent critical data traffic without encryption, default: prefer
  * Script to generate configure using autoconf/reconf and execute it

 -- William Grzybowski <william@agencialivre.com.br>  Tue, 24 Apr 2009 09:10:00 -0300

pam-pgsql (0.6.6) experimental; urgency=low

  * Removed references for debian/ package

 -- William Grzybowski <william@agencialivre.com.br>  Tue, 24 Mar 2009 09:47:50 -0300

pam-pgsql (0.6.5) experimental; urgency=low

  * Applied session management patch (Thanks to Olivier Thauvin)
  * Added missing include in pam_get_service.c
  * Added RPM spec (Thanks to Jose Arthur Benetasso Villanova)
  * Added ability to return PAM_PERM_DENIED

 -- Primoz Bratanic <primoz@slo-tech.com>  Fri, 29 Aug 2008 14:02:04 +0200

pam-pgsql (0.6.4) experimental; urgency=high

  * Fixed http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=481970
  * Critical security update, please upgrade

 -- Primoz Bratanic <primoz@slo-tech.com>  Sun, 25 May 2008 01:43:36 +0200

pam-pgsql (0.6.3) experimental; urgency=low

  * typo (stlen)
  * reupload

 -- Primoz Bratanic <primoz@slo-tech.com>  Wed, 17 May 2006 00:03:06 +0200

pam-pgsql (0.6.2) experimental; urgency=low

  * fixed memory leak with crypt function
  * fixed segfault with option "expired" set

 -- Primoz Bratanic <primoz@slo-tech.com>  Wed, 17 May 2006 00:01:28 +0200

pam-pgsql (0.6.1) experimental; urgency=low

  * Added ability to reference RHOST IP in queries
  * Changed configure.in to check for proper version of libpq
  * Fixed one more memory leak
  * Fixed type which prevented skiping acct part when acct is not configured

 -- Primoz Bratanic <primoz@slo-tech.com>  Wed,  1 Jun 2005 19:21:14 +0200

pam-pgsql (0.6) experimental; urgency=low

  * Added sanity checks (now we handle the malloc failure)
  * Removed call to free_module_options before *options has even been allocated (or set to NULL)
  * Removed obsolete TODO file
  * Removed obsolete public_key folder of previous maintainer
  * Fixed some typos in documentation
  * Added "authtok" option as an alternative to use_first_pass (to have same interface as pam_unix)
  * Removed "oldauthtok*" options (they have no use)
  * Removed "authtok_*" options (reverted to previous use_first_pass/try_first_pass)
  * Fixed try/use_first_pass not to require oldauthtok (so stacking works for passwd mode)
  * Deleted duplicated encrypt password code
  * Removed special behaviour of std_flags (they are in options now)
  * Major cleanup of pam_sm_authenticate
  * Removed unneeded db connection in pam_sm_chauthtok
  * Added new option to specify complete pgsql connection string (deprecates previous partial solutions)
  * Added auth_query, acct_query, pwd_query (obsoletes a lot of previous partial solutions)
  * Moved to PQexecParams (obsoletes need for data escaping)
  * Removed multiple calls to pam_set_item (duplicated)
  * Fixed all compile-time warnings
  * Removed all explicit escaping
  * Added -fstrict-aliasing -O2
  * New config.guess, config.sub
  * Respect PAM_DISALLOW_NULL_AUTHTOK in auth
  * Acct returns "require new password" with PAM_DISALLOW_NULL_AUTHTOK and null auth tok
  * Allowed additional logging query for failed or successful authentication
  * Major cleanup of pam_sm_chauthtok
  * Configuration now describes new options (see README)

 -- Primoz Bratanic <primoz@slo-tech.com>  Wed,  1 Jun 2005 09:54:38 +0200

pam-pgsql (0.5.2-9) unstable; urgency=low

  * Reapplied security patches (Closes: #230875,#307784)
  * Boolean values works with boolean type as well (Closes: #130496)
  * Documentation typo (Closes: #218291)
  * Reapplied other NMU patches (Closes: #307366)
  * Allow port specification (Closes: #247536)
  * Reapplied "Stack-Friendly patch" (Closes: #139473)
  * Deleted wrong README.Debian (Closes: #204181)
  * Documented host and port options (Closes: #204439)
  * Reapplied patch to allow different config files (Closes: #236484)
  * Reapplied patch to support another MD5 type passwords (Closes: #142889)
  * Change "must change password" field (if any) to false after changing password
  * Deleted build-all from root (Closes: #240823)
  * Fixed few memory leaks (Closes: #280774)

backend_pgsql.c
/*
 * PAM authentication module for PostgreSQL
 * 
 * Based in part on pam_unix.c of FreeBSD. See COPYRIGHT
 * for licensing details.
 *
 * David D.W. Downey ("pgpkeys") <david-downey@codecastle.com> et al. (see COPYRIGHT)
 * William Grzybowski <william@agencialivre.com.br>
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <syslog.h>
#include <ctype.h>
#include <mhash.h>
#include <time.h>
#include <sys/time.h>
#include <libpq-fe.h>
#include <crypt.h>
#include <sys/types.h>
#include <unistd.h>
#include <netdb.h>

#include "backend_pgsql.h"
#include "pam_pgsql.h"

/* very private: used only in get_module_options */
char *
build_connect_string(struct module_options *options)
{
    int len;
    char *str;

    len = strlen(options->database) + strlen(" dbname=") 
	+ (options->db_host ? (strlen(options->db_host) + strlen(" host=")) : 0)
	+ (options->db_port ? (strlen(options->db_port) + strlen(" port=")) : 0)
	+ (options->db_timeout ? (strlen(options->db_timeout) + strlen(" connect_timeout=")) : 0) 
	+ (options->db_user ? (strlen(options->db_user) + strlen(" user=")) : 0)
	+ (options->db_password ? (strlen(options->db_password) + strlen(" password=")) : 0)
	+ 1;

    str = (char *) malloc(len);
    memset(str, 0, len);

    if(str == NULL)
	return NULL;
	
    /* SAFE */
    strncat(str, "dbname=", strlen("dbname="));
    strncat(str, options->database, strlen(options->database));
    if(options->db_host) {
	strncat(str, " host=", strlen(" host="));
	strncat(str, options->db_host, strlen(options->db_host));
    }
    if(options->db_port) {
	strncat(str, " port=", strlen(" port="));
	strncat(str, options->db_port, strlen(options->db_port));
    }    
    if(options->db_timeout) {
	strncat(str, " connect_timeout=", strlen(" connect_timeout="));
	strncat(str, options->db_timeout, strlen(options->db_timeout));
    }
    if(options->db_user) {
	strncat(str, " user=", strlen(" user="));
	strncat(str, options->db_user, strlen(options->db_user));
    }
    if(options->db_password) {
	strncat(str, " password=", strlen(" password="));
	strncat(str, options->db_password, strlen(options->db_password));
    }
    return str;
}

int
options_valid(struct module_options *options)
{
	if (options->pg_conn_str == 0 || options->auth_query == 0) {
		SYSLOG("the database connection string and auth_query options are required.");
		return PAM_AUTH_ERR;
	}
	return PAM_SUCCESS;
}

char *
build_conninfo(modopt_t *options)
{
    char *str;

	 if(options == NULL)
		 return NULL;

	 str = (char *) malloc(sizeof(char)*512);
    memset(str, 0, 512);

    /* SAFE */
    strncat(str, "dbname=", strlen("dbname="));
    strncat(str, options->db, strlen(options->db));

	if(options->host) {
		strncat(str, " host=", strlen(" host="));
		strncat(str, options->host, strlen(options->host));
	}
	if(options->port) {
		strncat(str, " port=", strlen(" port="));
		strncat(str, options->port, strlen(options->port));
	}    
	if(options->timeout) {
		strncat(str, " connect_timeout=", strlen(" connect_timeout="));
		strncat(str, options->timeout, strlen(options->timeout));
	}
	if(options->user) {
		strncat(str, " user=", strlen(" user="));
		strncat(str, options->user, strlen(options->user));
	}
	if(options->passwd) {
		strncat(str, " password=", strlen(" password="));
		strncat(str, options->passwd, strlen(options->passwd));
	}
	if(options->sslmode) {
		strncat(str, " sslmode=", strlen(" sslmode="));
		strncat(str, options->sslmode, strlen(options->sslmode));
	}

	return str;
}

/* private: open connection to PostgreSQL */
PGconn *
pg_connect(struct module_options *options)
{
	PGconn *conn;
	conn = PQconnectdb(options->pg_conn_str);
	if(PQstatus(conn) != CONNECTION_OK) {
		SYSLOG("PostgreSQL connection failed: '%s'", PQerrorMessage(conn));
		return NULL;
	}
	return conn;
}

PGconn *
db_connect(modopt_t *options)
{
	PGconn *conn;
	char *conninfo;
	conninfo = build_conninfo(options);
	conn = PQconnectdb(conninfo);
	if(PQstatus(conn) != CONNECTION_OK) {
		SYSLOG("PostgreSQL connection failed: '%s'", PQerrorMessage(conn));
		return NULL;
	}
	return conn;
}

/* private: expand query; partially stolen from mailutils */

int
expand_query (char **command, const char** values, const char *query, const char *service, const char *user, const char *passwd, const char *rhost, const char *raddr)
{
	char *p, *q, *res;
	unsigned int len;
	unsigned int nparm=0;
  
	if (!query) {
		*command = NULL;
		return 0;
	}
	/* Compute resulting query length */
	for (len = 0, p = (char *) query; *p; ) {
		if (*p == '%') {
			if (p[1] == 'u' || p[1] == 'p' || p[1] == 's') {
				len += 4; /*we allow 128 tokens max*/
				p += 2;
				continue;
			} else if (p[1] == '%') {
				len++;
				p += 2;
				continue;
			}
		}
		len++;
		p++;  
	}
	res = malloc (len + 1);
	if (!res) {
		*command = NULL;
		return 0;
	}
	for (p = (char *) query, q = res; *p; ) {
		if (*p == '%') {
			switch (*++p) {
				case 'u': {
					sprintf(q, "$%i", ++nparm);
					values[nparm-1] = user;
					q += strlen (q);
					p++;
				}
				break;
				case 'p': {
					sprintf(q, "$%i", ++nparm);
					values[nparm-1] = passwd;
					q += strlen (q);
					p++;
				}
				break;
				case 's': {
					sprintf(q, "$%i", ++nparm);
					values[nparm-1] = service;
					q += strlen (q);
					p++;
				}
				break;
				case 'h': {
					sprintf(q, "$%i", ++nparm);
					values[nparm-1] = rhost;
					q += strlen (q);
					p++;
				}
				break;
				case 'i': {
					sprintf(q, "$%i", ++nparm);
					values[nparm-1] = raddr;
					q += strlen (q);
					p++;
					if (!raddr) {
						if (strchr(rhost, '.') != NULL) {
							*command = NULL;
							free (res);
							return 0;
						}
					}
				}
				break;
				case '%':
				default:
					*q++ = *p++;
				break;
			}
		} else	*q++ = *p++;
	 }
	 *q = 0;
	 
	 *command = res;
	 values[nparm] = NULL; 
	 return nparm;
}

/* private: execute query */
int
pg_execParam(PGconn *conn, PGresult **res, 
        const char *query, const char *service, const char *user, const char *passwd, const char *rhost)
{
	int nparm = 0;
	const char *values[128];
	char *command, *raddr;
	struct hostent *hentry;

	if (!conn) 
		return PAM_AUTHINFO_UNAVAIL;
	bzero(values, sizeof(*values));
	
	raddr = NULL;
	
	if(rhost != NULL && (hentry = gethostbyname(rhost)) != NULL) {
		/* Make IP string */
		raddr = malloc(16);
		sprintf(raddr, "%d.%d.%d.%d",
			hentry->h_addr_list[0][0],
			hentry->h_addr_list[0][1],
			hentry->h_addr_list[0][2],
			hentry->h_addr_list[0][3]);
		raddr[15] = 0;
	}
	
	nparm = expand_query(&command, values, query, service, user, passwd, rhost, raddr);
	if (command == NULL) 
		return PAM_AUTH_ERR;
	
	*res = PQexecParams(conn, command, nparm, 0, values, 0, 0, 0);
	free (command);
	free (raddr);
    
	if(PQresultStatus(*res) != PGRES_COMMAND_OK && PQresultStatus(*res) != PGRES_TUPLES_OK) {
		SYSLOG("PostgreSQL query failed: '%s'", PQresultErrorMessage(*res));
		return PAM_AUTHINFO_UNAVAIL;
	}
	return PAM_SUCCESS;
}

/* private: convert an integer to a radix 64 character */
static int
i64c(int i)
{
	if (i <= 0)
		return ('.');
	if (i == 1)
		return ('/');
	if (i >= 2 && i < 12)
		return ('0' - 2 + i);
	if (i >= 12 && i < 38)
		return ('A' - 12 + i);
	if (i >= 38 && i < 63)
		return ('a' - 38 + i);
	return ('z');
}

/* authenticate user and passwd against database */
int
auth_verify_password(const char *service, const char *user, const char *passwd, const char *rhost,
                     struct module_options *options)
{
	PGresult *res;
	PGconn *conn;
	int rc;
	char *tmp;

	if(!(conn = pg_connect(options)))
		return PAM_AUTH_ERR;

	DBGLOG("query: %s", options->auth_query);
	rc = PAM_AUTH_ERR;	
	if(pg_execParam(conn, &res, options->auth_query, service, user, passwd, rhost) == PAM_SUCCESS) {
		if(PQntuples(res) == 0) {
			rc = PAM_USER_UNKNOWN;
		} else {
			char *stored_pw = PQgetvalue(res, 0, 0);
			if (!strcmp(stored_pw, (tmp = encrypt_password(options, passwd, stored_pw)))) rc = PAM_SUCCESS; 
			free (tmp);
		}
		PQclear(res);
	}
	PQfinish(conn);
	return rc;
}


int
backend_authenticate(const char *service, const char *user, const char *passwd, const char *rhost, modopt_t *options)
{
	PGresult *res;
	PGconn *conn;
	int rc;
	char *tmp;

	if(!(conn = db_connect(options)))
		return PAM_AUTH_ERR;

	DBGLOG("query: %s", options->query_auth);
	rc = PAM_AUTH_ERR;	
	if(pg_execParam(conn, &res, options->query_auth, service, user, passwd, rhost) == PAM_SUCCESS) {
		if(PQntuples(res) == 0) {
			rc = PAM_USER_UNKNOWN;
		} else {
			char *stored_pw = PQgetvalue(res, 0, 0);
			if (!strcmp(stored_pw, (tmp = password_encrypt(options, passwd, stored_pw)))) rc = PAM_SUCCESS; 
			free (tmp);
		}
		PQclear(res);
	}
	PQfinish(conn);
	return rc;
}

/* private: encrypt password using the preferred encryption scheme */
char *
encrypt_password(struct module_options *options, const char *pass, const char *salt)
{
	char *s = NULL;

	switch(options->pw_type) {
		case PW_CRYPT:
		case PW_CRYPT_MD5:
			if (salt==NULL) {
				s = strdup(crypt(pass, crypt_make_salt(options)));
			} else {
				s = strdup(crypt(pass, salt));
			}
		break;
		case PW_MD5: {
			char *buf;
			int buf_size;
			MHASH handle;
			unsigned char *hash;
			handle = mhash_init(MHASH_MD5);
			if(handle == MHASH_FAILED) {
				SYSLOG("could not initialize mhash library!");
			} else {
				unsigned int i;
				mhash(handle, pass, strlen(pass));
				hash = mhash_end(handle);
				if (hash != NULL) {
					buf_size = (mhash_get_block_size(MHASH_MD5) * 2)+1;
					buf = (char *)malloc(buf_size);
					bzero(buf, buf_size);

					for(i = 0; i < mhash_get_block_size(MHASH_MD5); i++) {
						sprintf(&buf[i * 2], "%.2x", hash[i]);
					}
					free(hash);
					s = buf;
				} else {
					s = strdup("!");
				}
			}
		}
		break;
		case PW_SHA1: {
			char *buf;
			int buf_size;
			MHASH handle;
			unsigned char *hash;
			handle = mhash_init(MHASH_SHA1);
			if(handle == MHASH_FAILED) {
				SYSLOG("could not initialize mhash library!");
			} else {
				unsigned int i;
				mhash(handle, pass, strlen(pass));
				hash = mhash_end(handle);
				if (hash != NULL) {
					buf_size = (mhash_get_block_size(MHASH_SHA1) * 2)+1;
					buf = (char *)malloc(buf_size);
					bzero(buf, buf_size);

					for(i = 0; i < mhash_get_block_size(MHASH_SHA1); i++) {
						sprintf(&buf[i * 2], "%.2x", hash[i]);
					}
					free(hash);
					s = buf;
				} else {
					s = strdup("!");
				}
			}
		}
		break;
		case PW_CLEAR:
		default:
			s = strdup(pass);
	}
	return s;
}

char *
password_encrypt(modopt_t *options, const char *pass, const char *salt)
{
	char *s = NULL;

	switch(options->pw_type) {
		case PW_CRYPT:
		case PW_CRYPT_MD5:
			if (salt==NULL) {
				s = strdup(crypt(pass, crypt_makesalt(options->pw_type)));
			} else {
				s = strdup(crypt(pass, salt));
			}
		break;
		case PW_MD5: {
			char *buf;
			int buf_size;
			MHASH handle;
			unsigned char *hash;
			handle = mhash_init(MHASH_MD5);
			if(handle == MHASH_FAILED) {
				SYSLOG("could not initialize mhash library!");
			} else {
				unsigned int i;
				mhash(handle, pass, strlen(pass));
				hash = mhash_end(handle);
				if (hash != NULL) {
					buf_size = (mhash_get_block_size(MHASH_MD5) * 2)+1;
					buf = (char *)malloc(buf_size);
					bzero(buf, buf_size);

					for(i = 0; i < mhash_get_block_size(MHASH_MD5); i++) {
						sprintf(&buf[i * 2], "%.2x", hash[i]);
					}
					free(hash);
					s = buf;
				} else {
					s = strdup("!");
				}
			}
		}
		break;
		case PW_SHA1: {
			char *buf;
			int buf_size;
			MHASH handle;
			unsigned char *hash;
			handle = mhash_init(MHASH_SHA1);
			if(handle == MHASH_FAILED) {
				SYSLOG("could not initialize mhash library!");
			} else {
				unsigned int i;
				mhash(handle, pass, strlen(pass));
				hash = mhash_end(handle);
				if (hash != NULL) {
					buf_size = (mhash_get_block_size(MHASH_SHA1) * 2)+1;
					buf = (char *)malloc(buf_size);
					bzero(buf, buf_size);

					for(i = 0; i < mhash_get_block_size(MHASH_SHA1); i++) {
						sprintf(&buf[i * 2], "%.2x", hash[i]);
					}
					free(hash);
					s = buf;
				} else {
					s = strdup("!");
				}
			}
		}
		break;
		case PW_CLEAR:
		default:
			s = strdup(pass);
	}
	return s;
}

char *
crypt_makesalt(pw_scheme scheme)
{
	static char result[12];
	int len,pos;
	struct timeval now;

	if(scheme==PW_CRYPT){
		len=2;
		pos=0;
	} else { /* PW_CRYPT_MD5 */
		strcpy(result,"$1$");
		len=11;
		pos=3;
	}
	gettimeofday(&now,NULL);
	srandom(now.tv_sec*10000+now.tv_usec/100+clock());
	while(pos<len)result[pos++]=i64c(random()&63);
	result[len]=0;
	return result;
}

char *
crypt_make_salt(struct module_options *options)
{
	static char result[12];
	int len,pos;
	struct timeval now;

	if(options->pw_type==PW_CRYPT){
		len=2;
		pos=0;
	} else { /* PW_CRYPT_MD5 */
		strcpy(result,"$1$");
		len=11;
		pos=3;
	}
	gettimeofday(&now,NULL);
	srandom(now.tv_sec*10000+now.tv_usec/100+clock());
	while(pos<len)result[pos++]=i64c(random()&63);
	result[len]=0;
	return result;
}
